### ISSUE TITLE
[KEYBOARD]: Can focus management be improved for Edit or Add panels that are added before their trigger button

### ISSUE THREAD

**Issue by 1Copenut** [usds/us-forms-system#320]
**Describe the bug**
Focus is not being moved from the Add (Something) button when users keypress or click. Keyboard and screen reader users depend on our applications to manage focus when we add things in the DOM, especially when we add them before the currently selected item. Screenshots attached below.

**To Reproduce**
1. Review a form that uses the [Array Field](https://github.com/usds/us-forms-system/blob/master/src/js/fields/ArrayField.jsx) component
2. Click on the Add Another... button at the bottom of the component
3. Verify the focus does not settle on the array container or heading (either would be good, heading would be ideal.)
4. Tab into the Array Field, until the Remove button has keyboard focus
5. Press the Remove button
6. Verify the Add Another... button does not receive keyboard focus again

**Expected behavior**
When users initiate an action to add a set of fields, their focus should be placed at the beginning of that new set of fields. Screen readers use the heading being read out as a cue that they have changed the page, and they are oriented to the item or items that have been changed.

**Screenshots**
![screen shot 2018-11-30 at 1 18 14 pm](https://user-images.githubusercontent.com/934879/49384193-19289f80-f6e0-11e8-9831-388568d0317a.png)

**Desktop (please complete the following information):**
 - OS: MacOS Mojave
 - Browser: Chrome latest

**Comment by dmethvin-gov** [comment]
Yep, sounds like [handleAdd()](https://github.com/usds/us-forms-system/blob/37c569781d188be421eed28e9b5987aeea3e6370/src/js/review/ArrayField.jsx#L102) needs to call `focusElement()` after it's done scrolling to the new element. Conversely, [handleRemove()](https://github.com/usds/us-forms-system/blob/37c569781d188be421eed28e9b5987aeea3e6370/src/js/review/ArrayField.jsx#L115) just scrolls to the top so in that case I guess we'd set focus to the first form element? I may have missed some nuances since this component has a split personality depending on whether you're in editing mode or not.

**Comment by 1Copenut** [comment]
@dmethvin-gov I fully agree with calling focus when the scroll event is completed. This should trigger screen readers to announce the new element(s) to users, and give keyboard users an immediate inroad to `TAB` and start entering data.

Setting focus on the Remove button is a bit trickier. Moving focus all the way back up to the top of the form was a bit confusing during screen reader testing. I didn't know where I was on the page, and had to tab through all the content again after I had re-oriented. There are form shortcuts for JAWS and NVDA that could shortcut the long path through, but I still wonder if there's a closer landmark we could target. 

If I'm reading your comment correctly, the Remove has two states Edit and Read. I'm guessing the button has the same effect, but there's no good way of knowing what elements are before or after the panel being removed.

Is there a programmatic way to determine the previous element that would not be removed, or the next element after those being removed? I feel this would be a better user experience for keyboard and screen reader users to keep focus in the general area of the item removed, if possible.

Happy to discuss this further in terms of UX, lib capability, etc.

**Comment by dmethvin-gov** [comment]
There is probably a way to determine the previous element. Right now having Remove scroll to the top I agree may not be the best place to go.

**Comment by rianfowler** [comment]
so here's what I have so far: 

## ArrayList on Review Page

<img width="680" alt="screen shot 2018-12-20 at 10 08 51" src="https://user-images.githubusercontent.com/4998130/50296101-45ae1c80-043f-11e9-947d-15d52d21574b.png">

- Append the section title to the [`Add Another`](https://github.com/usds/us-forms-system/blob/37c569781d188be421eed28e9b5987aeea3e6370/src/js/review/ArrayField.jsx#L190) button e.g. `Add Another Service Period`
- Move the [`Add Another`](https://github.com/usds/us-forms-system/blob/37c569781d188be421eed28e9b5987aeea3e6370/src/js/review/ArrayField.jsx#L190) button to the bottom of the array list. 
- Add set focus behavior to [`handleAdd`](https://github.com/usds/us-forms-system/blob/37c569781d188be421eed28e9b5987aeea3e6370/src/js/review/ArrayField.jsx#L102) that focuses the `<h5>New Item Name</h5>`

## ArrayList Item- Remove

- Add an `aria-label` with the [itemName](https://github.com/usds/us-forms-system/blob/37c569781d188be421eed28e9b5987aeea3e6370/src/js/review/ArrayField.jsx#L208) to the [Edit button](https://github.com/usds/us-forms-system/blob/95c6518d5a0cc3a40f7376840760026b16007e13/src/js/review/ObjectField.jsx#L124) on the ObjectField
- Add set focus behavior to [`handleRemove`](https://github.com/usds/us-forms-system/blob/37c569781d188be421eed28e9b5987aeea3e6370/src/js/review/ArrayField.jsx#L115) to set focus to 
  - when 0 items after the removed item: the add another button
  - when 1 or more items after the removed item: the next item in the list i.e. the item that takes the index of the removed item 

Questions: 
- Regarding the last item above: Do you think it might be weird to "Remove Service Period 3" and then set focus to "Service Period 3"
- What do you think about pushing the confirmation step into a separate issue? might be useful to do it after we're done with this so we're not trying to iterate on too many things at once.

**Comment by 1Copenut** [comment]
@rianfowler In regard to your questions above, I think the right approach after you click "Remove Service Period 3" might be to set focus on "Service Period 2", the heading or container. That way something tangible is read out to screen readers. Would there ever be a situation where there are no service periods or ArrayLists on the page? Trying to think of edge cases like users remove the last one, and have no data fields to input to.

I think it would be nice to have a confirmation step for removing previously entered data, like a modal asking "Do you want to remove this?". IMO, that would benefit from being tracked as a separate issue.

**Comment by rianfowler** [comment]
>Would there ever be a situation where there are no service periods or ArrayLists on the page?
I think in this scenario we would just focus on the add button. 

But you're thinking of situations where we're showing the list because of another question like **Do you have any dependents?** shows the array list and what happens if they want to remove everything from the list? do we focus on the question? 

I'll make another ticket for the confirmation thing.

**Comment by 1Copenut** [comment]
Yeah, I think so. The implementations I've seen so far always have at least one `ArrayList` component that can't be removed from the page, with the possibility to add others. I might be inventing edge cases; if we're always going to have at least one `ArrayList` then there would never be a case where pressing Remove had to set focus anywhere but the previous component heading.

**Comment by rianfowler** [comment]
@1Copenut @dmethvin-gov If there aren't any objections, I'm going to take a stab at this on the 1.1 branch. Trevor I think I'll make a WIP on `vets-website` that we can test with locally this week- is that something you can do with your accessibility tools or is it easier to do it on a review instance?

**Comment by 1Copenut** [comment]
@rianfowler I can test this on a `vets-website` feature branch. Since this is so heavily focused on keyboard focus management, it shouldn't require any tools than a keyboard and a test user account.

**Comment by dmethvin-gov** [comment]
Fixed in the 1.x-stable branch by #332
